### PR TITLE
Feat: Added the ability to edit the configuration file

### PR DESCRIPTION
**Pull request by Trico-Everfire**
This is a long commit with a lot of changes.
I added the ability to edit the layout and items inside the configuration file
from within the launcher itself. I also documented everything, even stuff done
before this, to make the code both more human readable and understandable.

### CODE REVIEW

File: src/editconfig.cpp [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+		}
+	}
+
+	// We set the general alignment for all the items in the configuration layout.
+	pEditConfigurationLayout->setAlignment( Qt::AlignTop );
+
+	// We then create the 5 buttons responsible for editing individual
+	// instances of executables/urls/categories.
+	pEditConfigurationLayout->addWidget( m_pEditList, 0, 0, 5, 1 );
+	auto addCurrentButton = new QPushButton( this );
```
Comment (AWildErin) [inline src/editconfig.cpp]: Stuff like this generally should be prefixed with `p`, they're pointers

File: src/editconfig.cpp [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+	m_pTypeLabel = new QLabel( this );
+	m_pTypeLabel->setText( "Type:" );
+	pConfigPopupLayout->addWidget( m_pTypeLabel, 0, 2 );
+	m_pTypeComboBox = new QComboBox( this );
+	m_pTypeComboBox->addItem( "Executable" );
+	m_pTypeComboBox->addItem( "Url" );
+	m_pTypeComboBox->addItem( "Category" );
+	pConfigPopupLayout->addWidget( m_pTypeComboBox, 0, 3 );
+
+	m_pUrlLabel = new QLabel( this );
```
Comment (AWildErin) [inline src/editconfig.cpp]: These don't really need to be members. You're not accessing them from anywhere outside this function?

File: src/editconfig.cpp [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+	{
+		// We first check if there is an item in the first place.
+		// Then check if the item is a category.
+		auto current = m_pEditList->item( 0 );
+		if ( !current )
+		{
+			m_pApplyButton->setDisabled( true );
+			m_pApplyButton->setToolTip( "The top element must be a category." );
+			return;
+		};
```
Comment (AWildErin) [inline src/editconfig.cpp]: there does not need to be `;`

**Review by AWildErin — CHANGES_REQUESTED** [review]
I'm going to be brutally honest on this PR: Ideally half of the comments here are useless. There is such thing as over-documenting, and this is sort of an example of that. 

I'm also not a fan of how verbose you made certain names be `onProcessPushButtonPressedCallback`, `newListItemJSONContents` etc. The config code is fine, but I'm just not a huge fan of how much documentation you added. Good code usually is self documenting, and besides this was started out by myself and generally was going to be edited by myself, so documentation wasn't really necessary.

File: src/editconfig.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+	// The popup for config items we want to edit.
+	class CEditConfigPopup : public QDialog
+	{
+		bool applyChanges = false;
+
+	public:
+		CEditConfigPopup( CEditConfig *parent );
+		bool shouldApplyChanges() const;
+		QComboBox *m_pTypeComboBox;
```
Comment (AWildErin) [inline src/editconfig.h:30]: Also, a small thing I just noticed, this stuff should probably be private with accessor funcs for them

File: src/editconfig.cpp (hunk 2 above)
Comment (Trico-Everfire) [reply src/editconfig.cpp]: There was a weird bug early in development that required them to be, it had to do with callback lambdas not accessing them, but I figured out the cause, yet... forgot about these... I'll fix it.

File: src/editconfig.cpp [hunk 5]
```diff
@@ -0,0 +1,559 @@
+#include "editconfig.h"
+
+#include <QDialogButtonBox>
+#include <QFileInfo>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QSizePolicy>
+
+ui::CEditConfig::CEditConfig( CMainView *parent ) :
```
Comment (AWildErin) [inline src/editconfig.cpp]: Ideally you should probably do `using namespace ui` below the includes and remove the `ui::` here

File: src/editconfig.cpp (hunk 1 above)
Comment (Trico-Everfire) [reply src/editconfig.cpp]: I know (it's why every other pointer has `p`, That one is like the 1 button I forgot when refactoring the code to be human readable.

**Comment by Trico-Everfire** [comment]
As to reply to your message, Even if it's supposed to be only used by you, having the documentation on what my code does can be handy, I'd rather over-document something than under-document, if you wanna get rid of any redundant documentation, feel free to point out what needs to be removed, because to me, I couldn't tell what documentation is needed and what isn't.
as for long variable names, I tried to make them as human readable as possible, Which lead to the long names. If you have alternative names for them I'd love to implement those, as I agree some are way too long.

File: src/editconfig.h (hunk 4 above)
Comment (Trico-Everfire) [reply src/editconfig.h:30]: so...
```c++
QComboBox *getTypeComboBox()
{
    return m_pTypeComboBox;
}
```
???

File: src/editconfig.cpp (hunk 5 above)
Comment (Trico-Everfire) [reply src/editconfig.cpp]: Fixed.

File: src/editconfig.cpp (hunk 3 above)
Comment (Trico-Everfire) [reply src/editconfig.cpp]: fixed.

File: src/editconfig.cpp (hunk 2 above)
Comment (Trico-Everfire) [reply src/editconfig.cpp]: Fixed.

File: src/editconfig.cpp (hunk 1 above)
Comment (Trico-Everfire) [reply src/editconfig.cpp]: Fixed.

File: src/editconfig.h (hunk 4 above)
Comment (AWildErin) [reply src/editconfig.h:30]: In fact I suppose this doesn't matter actually, so don't need to bother with this.

Comment (Trico-Everfire) [reply src/editconfig.h:30]: Okay.

**Comment by Trico-Everfire** [comment]
There's currently a segfaul;t, I'm working on fixing it.

**Comment by Trico-Everfire** [comment]
Segfault fixed.

File: src/config.h [hunk 6]
```diff
@@ -15,6 +15,8 @@
 #define OS_BIN_SUFFIX ".sh"
 #endif
 
+// We define a struct with the required
```
Comment (ENDERZOMBI102) [inline src/config.h:18]: prob overdoc'd here

File: src/editconfig.cpp [hunk 7]
```diff
@@ -0,0 +1,560 @@
+#include "editconfig.h"
```
Comment (ENDERZOMBI102) [inline src/editconfig.cpp:1]: so many comments...  would betted be less imo

**Review by ENDERZOMBI102 — COMMENTED** [review]
code seems pretty ok

File: src/editconfig.cpp (hunk 7 above)
Comment (Trico-Everfire) [reply src/editconfig.cpp:1]: Can you specify what's being overdocumented here? I wanna know what I can keep and toss out.

**Comment by Trico-Everfire** [comment]
oops, didn't meant to reassign...

**Review by AWildErin — APPROVED** [review]

**Pull request merged**